File: matlab-version/Yoshino_olivine_conductivity_Xfe.m
function sigma = Yoshino_olivine_conductivity_Xfe(temperature, water_content,pressure,Xfe)
    % Constants
    % This function calculates the electrical conductivity based on the
    % provided temperature (T), water content (Cw), and material constants.
    % Constants
    k = 1.380649*10^(-23);
    J2eV = 1.0/(1.60217733*10^(-19)); % 1eV = 1.60217733*10^(-19) J
    k = k*J2eV;
    
    sigma_i = 10^4.73;
    Hi = 2.31;
    
    sigma_h = 525;%-15;
    Hh = 1.96;% + 0.01;
    alpha_h = 1.49;
    V0 = -0.01;
    belta_h = 0.22;
    
    sigma_p = 10^1.9;
    Hp = 0.92;
    alpha = 0.16;
    
    T =  temperature;
    Cw = water_content;     
    % Check if temperature is greater than zero
    if any(temperature <= 0)
        error('Temperature must be greater than zero Kelvin.');
    end

    % Ensure pressure is numeric
    if iscell(pressure)
        pressure = cell2mat(pressure); % Convert cell array to numeric if necessary
    end

    % Check if pressure is non-negative
    if any(pressure < 0)
        error('Pressure must be non-negative.');
    end

    % Convert pressure from Pa to J/cc
%     pressure = 1e-6 * pressure;  % 1 Pa = 1 J/m³ = 1e-6 J/cc

    % Ensure water_content is numeric
    if iscell(water_content)
        water_content = cell2mat(water_content); % Convert cell array to numeric if necessary
    end

    % Check if water_content is within the range of 0 to 1
    if any(water_content < 0) || any(water_content > 1)
        error('Water content must be in the range of 0 to 1.');
    end

    % Sum the conductivities
    sigma = sigma_i*exp(-Hi/(k*T))+...
        sigma_h*Xfe*exp(-((Hh-alpha_h*Xfe^(1.0/3.0) + 0.0104*pressure*(V0-belta_h*Xfe)))/(k*T))+...
        sigma_p*Cw*exp(-(Hp-alpha*Cw^(1.0/3.0))/(k*T));
end

File: cpp-version/Yoshino_olivine_conductivity_Xfe.cpp
#include <cmath>
#include <stdexcept>
#include <iostream>

double Yoshino_olivine_conductivity_Xfe(double temperature, double water_content, double pressure, double Xfe) {
    // Constants
    const double k_boltzmann = 1.380649e-23;  // Boltzmann constant in J/K
    const double J2eV = 1.0 / 1.60217733e-19; // 1 eV = 1.60217733e-19 J
    const double k = k_boltzmann * J2eV;      // Boltzmann constant in eV/K
    const double sigma_i = std::pow(10, 4.73); // Intrinsic conductivity in S/m
    const double Hi = 2.31;                    // Activation enthalpy for intrinsic conduction in eV
    const double sigma_h = 525;                // Hydrogen-related conductivity in S/m
    const double Hh = 1.96;                    // Activation enthalpy for hydrogen in eV
    const double alpha_h = 1.49;               // Constant for hydrogen conduction
    const double V0 = -0.01;                   // Activation volume
    const double belta_h = 0.22;               // Beta constant for hydrogen
    const double sigma_p = std::pow(10, 1.9);  // Proton-related conductivity in S/m
    const double Hp = 0.92;                    // Activation enthalpy for proton conduction in eV
    const double alpha = 0.16;                 // Constant for proton conduction

    // Input validation
    if (temperature <= 0) {
        throw std::invalid_argument("Temperature must be greater than zero Kelvin.");
    }

    if (pressure < 0) {
        throw std::invalid_argument("Pressure must be non-negative.");
    }

    if (water_content < 0 || water_content > 1) {
        throw std::invalid_argument("Water content must be in the range of 0 to 1.");
    }

    if (Xfe < 0 || Xfe > 1) {
        throw std::invalid_argument("Iron content (Xfe) must be in the range of 0 to 1.");
    }

    // Sum the conductivities
    double sigma = sigma_i * std::exp(-Hi / (k * temperature)) +
                   sigma_h * Xfe * std::exp(-((Hh - alpha_h * std::pow(Xfe, 1.0 / 3.0) + 0.0104 * pressure * (V0 - belta_h * Xfe))) / (k * temperature)) +
                   sigma_p * water_content * std::exp(-(Hp - alpha * std::pow(water_content, 1.0 / 3.0)) / (k * temperature));

    return sigma;
}

// Example usage
int main() {
    double temperature = 1000.0;  // Example temperature in Kelvin
    double water_content = 0.1;   // Example water content
    double pressure = 1e5;        // Example pressure in Pascal
    double Xfe = 0.1;             // Example iron content fraction

    try {
        double conductivity = Yoshino_olivine_conductivity_Xfe(temperature, water_content, pressure, Xfe);
        std::cout << "Conductivity: " << conductivity << " S/m" << std::endl;
    } catch (const std::invalid_argument& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }

    return 0;
}
